File: rust/src/problems/p206.rs
use crate::utils::linked_list::ListNode;

pub struct Solution;

impl Solution {
    pub fn reverse_list(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut prev = None;
        while let Some(mut node) = head {
            head = node.next;
            node.next = prev;
            prev = Some(node);
        }
        prev
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        let l1 = ListNode::from_vec(&[1, 2, 4]);
        let result = Solution::reverse_list(l1);
        let expected = ListNode::from_vec(&[4, 2, 1]).unwrap().to_string();
        assert_eq!(expected, result.unwrap().to_string());
    }
}

File: cpp/p206.cpp
#include "helpers.h"
#include <cassert>

class Solution {
public:
    ListNode *reverseList(ListNode *head) {
        ListNode *prev = nullptr;
        while (head) {
            auto next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
};

int main() {
    auto head = parse_singly_linked_list({1, 2, 3, 4, 5});
    auto expected = parse_singly_linked_list({5, 4, 3, 2, 1});

    auto r = Solution().reverseList(head);

    while (expected) {
        assert(r && expected->val == r->val);
        r = r->next;
        expected = expected->next;
    }
    assert(!r);
}
